sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,190 @@
+             [-1, 2, -3, -0.5], \
+             [ 0, 3, -2,  0.5], \
+             [ 1, 4, -1,    2]]
+    >>> dis = KBinsDiscretizer(n_bins=3)
+    >>> dis.fit(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    KBinsDiscretizer(...)
+    >>> dis.cut_points_
+    array([[-1.,  2., -3.,  0.],
+           [ 0.,  3., -2.,  1.]])
+    >>> dis.transform(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    array([[-2., 1., -4., -1.],
+           [-1., 2., -3., -1.],
+           [ 0., 3., -2.,  0.],
+           [ 0., 3., -2.,  1.]])
+    """
+
+    sparse_formats = ['csr', 'csc']
+
+
+    def __init__(self, n_bins=2):
+        self.copy = copy
+        self.n_bins = n_bins
+
+        # Attributes
+        self.min_ = None
+        self.max_ = None
+        self.n_features_ = None
+
+
+    def fit(self, X, y=None):
+        """Finds the intervals of interest from the input data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse-matrix}, shape [n_samples, n_features]
+            The input data containing continuous features to be
+            discretized. Categorical columns are indicated by
+            categorical_columns from the constructor.
+
+        Returns
+        -------
+        self
+        """
+
+        if self.n_bins < 2:
+            raise ValueError("Discretizer received an invalid number "
+                             "of bins. Received {0}, expected at least 2."
+                             .format(self.n_bins))
+
+        X = check_array(X, accept_sparse=self.sparse_formats, dtype=float,
+                        ensure_2d=False)
+        self.n_features_ = X.shape[1]
+
+        self.min_ = X.min(axis=0)
+        self.max_ = X.max(axis=0)
+        return self
+
+
+    def transform(self, X, y=None):
+        """Discretizes the input data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse-matrix}, shape [n_samples, n_features]
+            The input data containing continuous features to be
+            discretized. Categorical columns are indicated by
+            categorical_columns from the constructor.
+
+        Returns
+        -------
+        T : array-like, shape [n_samples, n_features]
+            Array containing discretized features alongside the
+            categorical features. Categorical features are represented
+            as the last features of the output.
+
+        """
+        check_is_fitted(self, ["min_", "max_"])
+        X = check_array(X, accept_sparse=self.sparse_formats, dtype=float,
+                        ensure_2d=False)
+
+        if X.shape[1] != self.n_features_:
+            raise ValueError("Transformed array does not have currect number "
+                             "of features. Expecting {0}, received {1}."
+                             .format(self.n_features_, X.shape[1]))
+
+        # TODO: Make copy parameter?
+        X = X.copy()
+
+        # Clip outliers from data (TODO)
+        #X[X > self.max_] = self.max_
```

Comment:
Well I'll be damned. [np.clip](https://docs.scipy.org/doc/numpy/reference/generated/numpy.clip.html#numpy.clip) exists.

This doesn't solve how we would adapt this solution to sparse matrices though.